tools/variant_simulator/simulate_variation.pl

```diff
@@ -0,0 +1,200 @@
+sub processGenes{
+  my $geneArr = shift;
+  my $fho = shift;
+
+  foreach my $geneID (@$geneArr) {
+    next unless $geneID->biotype eq $biotype;
+
+    my @parts; # set of exons OR introns
+    if ($codingOnly){
+      my %exons;
+      my @transc = @{ $geneID->get_all_Transcripts() };
+      while ( my $transcript = shift @transc ) {
+        next unless $transcript->biotype() eq $biotype;
+        foreach my $exon (@{ $transcript->get_all_translateable_Exons() }) {
+            $exons{$exon->stable_id} = $exon;
+        }
+      }
+      @parts = values %exons;
+    } elsif ($exonOnly){
+      @parts = @{$geneID->get_all_Exons()};
+    } elsif($intronOnly){
+      @parts = @{$geneID->get_all_Introns()};
+    } else {
+      @parts = ($geneID);
+    }
+
+    generateSNPs(\@parts,$geneID->external_name, $fho);
+  }
+}  
+
+#subroutine: for a given set of features + given edge, write out all possible SNPs
+sub generateSNPs {
+  my $elements = shift;
+  my $gene_name = shift;
+  my $fho = shift;
+
+  my @bases = qw/A C T G/;
+  while( my $part = shift (@$elements)){
+    my $start = $part->seq_region_start - $edges;
+    my $pos = $start;
+    my $chrom = $part->seq_region_name;
+    my $partID;
+    $partID = $part->stable_id if $exonOnly;
+    $partID = $part->display_id if $intronOnly;
+    $partID  = $part->display_id;
+    my $seq = $slice_adaptor->fetch_by_region("chromosome", $chrom, $start ,$part->seq_region_end + $edges)->seq;
+    my @seqAr = split("", $seq);
+    my ($id, $qual, $filter, $info) = (".", ".", ".", ".");
+    for ( my $i = 0; $i < @seqAr; $i++ ) {
+         my $ref_seq = $seqAr[$i];
+         my @tmp_alts = grep {$_ ne $ref_seq} @bases;
+         $pos = $start + $i;
+         $info = join(":", ($gene_name, $partID));
+         while(my $tmp_alt = shift @tmp_alts) {
+           $id = $chrom."-".$pos."-".$ref_seq."-".$tmp_alt;
+           printf $fho "%s\t%s\t%s\t%s\t%s\t", $chrom, $pos, $id, $ref_seq, $tmp_alt;
+           printf $fho "%s\t%s\t%s\n", $qual, $filter, $info;
+        }
+    }
+  }
+}
+
+sub usage {
+
+  print qq{
+  Usage: perl simulate_variation.pl [OPTIONS]
+
+  Generate a VCFv4 file with all possibel SNPs for protein_coding genes given specific species or chromosome or gene.
+
+  Options:
+
+      -registry          Ensembl registry file containing database connection details (Required)
+      -species           Progress information is printed (Default: human)
+
+      -output|o          Output file (Default: simulated.vcf)
+
+      -chrom             Generate SNPs only for specified chromosome
+      -gene              Generate all SNPs for specified gene
+      -codingOnly        Generate SNPs only for translatable exons of protein_coding transcripts
+      -exonsOnly         Generate SNPs only for all exons of the genes
+      -intronsOnly       Generate SNPs only for all introns of the genes
+
+      -edge              Upstream/downstream bp of each feature (Default: 0)
```

Comment:
Can this description be clarified? 'Length of flanking region in which to generate SNPs' ?